repos/system_upgrade/el7toel8/actors/openldapnsstoopenssl/actor.py

```diff
@@ -0,0 +1,58 @@
+# vim: tabstop=8 expandtab shiftwidth=4 softtabstop=4
+from leapp.actors import Actor
+from leapp.libraries.actor.openldap import process as ol_process
+from leapp.libraries.actor.sssd import process as sd_process
+from leapp import reporting
+from leapp.reporting import Report, create_report
+from leapp.tags import ApplicationsPhaseTag, IPUWorkflowTag
+
+
+class OpenldapNSStoOpenSSL(Actor):
+    """
+    Convert old NSS-style TLS configuration to the new OpenSSL-style configuration.
+
+    Configuration in /etc/openldap/ldap.conf will be converted into
+    OpenSSL-style configuration if necessary. The converted certificates
+    will be placed into /etc/openldap/certs-migrated and ldap.conf will be
+    updated with configuration pointing to extracted certificates. Similar
+    process will occur for SSSD where certificates configuration is expected.
+    """
+    name = 'openldap_nss_to_openssl'
+    produces = (Report,)
+    tags = (ApplicationsPhaseTag, IPUWorkflowTag)
+
+    def process(self):
+        if not self.handle_errors(ol_process, self.abandon, self.fail):
+            self.fail('OpenLDAP certificates were not properly converted.')
+        if not self.handle_errors(sd_process, self.abandon, self.fail):
+            self.fail('SSSD configuration of OpenLDAP certificates was not properly converted.')
+
+    def handle_errors(self, process, abandon, fail):
+        rv, err = process(self.log)
+        if rv is None:
+            abandon(err)
+            return False
+        elif rv is False:
+            fail(err)
+            return False
+        elif rv is True:
+            self.success()
+            return True
+        fail('Unexpected error occurred: %s' % err)
+        return False
+
+    def abandon(self, reason):
+        create_report([reporting.Title('No OpenLDAP certificates conversion necessary'),
+                       reporting.Summary('%s. Certificates are most likely in correct format.'
+                                         ' Leaving as is.' % reason),
+                       reporting.Severity(reporting.Severity.LOW)])
+
+    def success(self):
+        create_report([reporting.Title('OpenLDAP certificates converted successfully'),
+                       reporting.Summary('Created PEM certificates and updated the configuration file.'),
+                       reporting.Severity(reporting.Severity.LOW)])
+
+    def fail(self, err):
+        create_report([reporting.Title('OpenLDAP certificates processing failed'),
+                       reporting.Summary(str(err)),
+                       reporting.Severity(reporting.Severity.HIGH)])
```

Comment:
Assuming this function is used to report that certain certificates couldn't be converted to the desired format this function should also enumerate them as `RelatedResources` i.e. :

```python
[reporting.RelatedResource('file', faulty_cert) for faulty_cert in faulty_certificates]
```

Reply:
@kenoh Yeah, we're on the right track here. To summarize - if the report entry relates to one or more resources on the system it should be spelled out so that the frontend(s) can pinpoint the correct resource and provide context-relevant information.

We currently distinguish the following resource types:

| Type | Semantics |
|-------|---------------|
| file | References a file on a filesystem |
| package | References an installable RPM |
| repository | References a RPM repository |
| kernel-module | References a specific kernel module |
| disk | References a disk or a volume |

For example:
```
reporting.RelatedResource('package', 'vim-enhanced')
```

will render as link to:
https://access.redhat.com/downloads/content/vim-enhanced/x86_64/package-latest
where the user can get more information about the package without having to search for it.

Reply:
Sounds nice, thanks for the update. I'll update the PR according to this.